clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -218,7 +248,13 @@ public RecordAppendResult append(TopicPartition tp,
                     // Somebody else found us a batch, return the one we waited for! Hopefully this doesn't happen often...
                     return appendResult;
                 }
-
+                if (manualPartition) {
+                    Integer part = partitioner.repartitionOnNewBatch(tp.topic(), key, value, cluster);
+                    if (part != null) {
+                        tp = new TopicPartition(tp.topic(), part);
+                        dq = getOrCreateDeque(tp);
```

Comment:
Does the locking here work? We are in a synchronized block of the old `dq` but here we re-assign the variable and modify it.

Reply:
Good catch. Would we want to close the block and open a new one on partition change? That makes sense to me.

Reply:
I agree we need to close the synchronization block here. We would need to open a new one and I think we might also want to check whether the new `dq` has an open batch

Reply:
I'm running through the logic here. So we close the old dq block since we don't have that partition anymore. When we open the new one, I'm wondering if there is the possibility for an open batch to exist. In ideal conditions, would it be possible to have other batches still open? Perhaps if two records triggered the repartitionOnNewBatch at the same time, but in that case I think we will always create two batches.